src/Compilers/CSharp/Portable/Binder/Binder_Patterns.cs

```diff
@@ -345,7 +345,8 @@ private ExpressionSyntax SkipParensAndNullSuppressions(ExpressionSyntax e)
                     HashSet<DiagnosticInfo>? useSiteDiagnostics = null;
                     if (expression.ConstantValue == ConstantValue.Null)
                     {
-                        if (inputType.IsNonNullableValueType())
+                        // Function pointers are value types, but they can be assigned null, so they can be matched against null.
+                        if (inputType.IsNonNullableValueType() && inputType.TypeKind != TypeKind.FunctionPointer)
```

Comment:
>inputType.TypeKind != TypeKind.FunctionPointer [](start = 66, length = 46)

It looks like we have the same issue with regular pointers, I think we should use IsPointerOrFunctionPointer helper here instead.